src/treiber_stack.mli

```diff
@@ -1,31 +1,191 @@
 (** Classic multi-producer multi-consumer Treiber stack.
 
-    All function are lockfree. It is the recommended starting point
-    when needing LIFO structure. *)
+  All functions are lock-free. It is the recommended starting point
+  when needing a LIFO structure. *)
+
+(** {1 API} *)
 
 type 'a t
-(** Type of Treiber stack holding items of type [t]. *)
+(** Represents a lock-free Treiber stack holding elements of type ['a]. *)
 
 val create : unit -> 'a t
-(** [create ()] returns a new and empty Treiber stack. *)
+(** [create ()] creates a new empty Treiber stack. *)
+
+val of_list : 'a list -> 'a t
+(** [of_list list] creates a new Treiber stack from a list. *)
 
 val is_empty : 'a t -> bool
-(** [is_empty s] checks whether stack [s] is empty. *)
+(** [is_empty stack] returns [true] if the [stack] is empty, otherwise [false]. *)
 
-val push : 'a t -> 'a -> unit
-(** [push s v] adds the element [v] at the top of stack [s]. *)
+(** {2 Consumer functions} *)
 
 exception Empty
-(** Raised when {!pop} is applied to an empty queue. *)
+(** Raised when {!pop_exn}, {!peek_exn}, {!drop_exn}, or {!set_exn} is 
+  applied to an empty stack.
 
-val pop : 'a t -> 'a
-(** [pop s] removes and returns the topmost element in the
-    stack [s].
+  This exception is meant to avoid allocations required by an option type.
+  As such, it does not register backtrace information and it is recommended to 
+  use the following pattern to catch it.
```

Comment:
Do we have examples of other libraries which do this? My understanding from [OCaml behind the scenes: Exception (around 47m12s)](https://watch.ocaml.org/w/vur2uhCFTZWQfDfkcbCPfC?start=47m12s) is that `raise Empty` is already allocation-free unless the developer has explicitly requested debugging backtraces (cc @fabbing in case I'm wrong ^^')

If maximum performances are required, would the following very verbose optin alternative API be allocation-free? (or would the cost of calling `default ()` be prohibitive?)

```ocaml
match pop_with s ~default:(fun () -> raise_notrace Empty) with
| value -> ...
| exception Empty -> ...
```